web/l10n.js

```diff
@@ -77,6 +80,22 @@ class L10n {
     }
   }
 
+  /** @inheritdoc */
+  async destroy() {
+    for (const element of this.#elements) {
+      this.#l10n.disconnectRoot(element);
+    }
+    this.#elements.clear();
+    this.#l10n.pauseObserving();
+
+    // Since `disconnectRoot`/`pauseObserving` can still trigger asynchronous
+    // operations, without any way to actually cancel them, attempt to
+    // workaround timing issues when closing the integration-tests.
+    await new Promise(resolve => {
```

Comment:
Interesting way to wait a little without an explicit timeout ;).

Reply:
This was prompted by reading https://github.com/projectfluent/fluent.js/blob/main/fluent-dom/src/dom_localization.js which I obviously forgot to mention in the commit message.

When pausing l10n Fluent will actually trigger translation of any pending elements via https://github.com/projectfluent/fluent.js/blob/fee2242049010f314345a7c3b7b91600df3a05fe/fluent-dom/src/dom_localization.js#L195-L200
That method then triggers the actual translation with a `requestAnimationFrame` in https://github.com/projectfluent/fluent.js/blob/fee2242049010f314345a7c3b7b91600df3a05fe/fluent-dom/src/dom_localization.js#L246-L256 and that one is only `null`ed but not actually cancelled in https://github.com/projectfluent/fluent.js/blob/fee2242049010f314345a7c3b7b91600df3a05fe/fluent-dom/src/dom_localization.js#L169-L175

Hence the idea in this patch, to utilize a "later" `requestAnimationFrame` in our code since that *should* guarantee that everything Fluent-related has actually finished.

Reply:
I'm not sure if this could perhaps be seen as a bug *in Fluent* since it doesn't call `cancelAnimationFrame(this.pendingrAF)`, or if the idea is that pending elements will *always* be translated even when calling `disconnectRoot`?

Reply:
@flodolo, @eemeli ^^

Reply:
I think it'd be nice to have a real `destroy` method in Fluent to remove the observer, cancel tasks, ...

Reply:
Indeed, for now I'm fine with doing it ourselves here given that this seems to resolve the shutdown issues, but having Fluent provide a `destroy` method would be even better.

Reply:
Note that #18318 is extra reason to have this implemented upstream in Fluent itself, because having to work around this here can lead to unexpected new issues.